### PR TITLE
Improve supports subjects validation SE-1981

### PR DESCRIPTION
**Pull request by peteryates**
### JIRA Ticket Number

SE-1981

### Context

When failing to enter a placement date, when the form is re-rendered the `supports_subjects` flag had been already set by `#create`.

### Changes proposed in this pull request

Now instead of setting the flag in the previous manner regardless of the
situation it's only set if validation has passed and there's a new
validation rule that ensures it's been set manually if the school
supports primary and secondary phases (ie has at least one phase that
does support subjects and one that doesn't)

Screenshot 2019-11-15 at 14 09 39

### Guidance to review

Do the changes look sensible? Could the new methods on `Bookings::School` be better named? They look ok to me without being too wordy, but perhaps I've overlooked something

### CODE REVIEW

File: app/models/bookings/placement_date.rb [hunk 1]
```diff
@@ -39,6 +39,13 @@ class PlacementDate < ApplicationRecord
       presence: true,
       on: :create
 
+    # users manually selecting this only happens when schools are both primary
+    # and secondary, otherwise it's automatically set in the controller
+    validates :supports_subjects,
+      inclusion: [true, false],
+      on: :create,
+      if: -> { bookings_school&.primary_and_secondary? }
```
Comment (rjlynch) [inline app/models/bookings/placement_date.rb]: do we need the `on: :create` + `if: -> { bookings_school&.primary_and_secondary? }` conditional here? Doesn't moving setting the `supports_subjects` flag in side the `valid?` conditional in the controller not handle the issue?

File: app/models/bookings/school.rb [hunk 2]
```diff
@@ -188,6 +188,14 @@ def supports_subjects?
     phases.any?(&:supports_subjects?)
   end
 
+  def does_not_support_subjects?
```
[the remark this answers is not in the document]
Comment (peteryates) [reply app/models/bookings/school.rb]: Agree

File: app/models/bookings/placement_date.rb (hunk 1 above)
Comment (peteryates) [reply app/models/bookings/placement_date.rb]: Also agree, good spot

File: app/models/bookings/placement_date.rb [hunk 3]
```diff
@@ -39,6 +39,12 @@ class PlacementDate < ApplicationRecord
       presence: true,
       on: :create
 
+    # users manually selecting this only happens when schools are both primary
+    # and secondary, otherwise it's automatically set in the controller
+    validates :supports_subjects,
+      inclusion: [true, false],
+      if: -> { bookings_school&.has_primary_and_secondary_phases? }
```
Comment (rjlynch) [inline app/models/bookings/placement_date.rb]: I don't think we need this if either!

Comment (peteryates) [reply app/models/bookings/placement_date.rb]: ah, thinking about it the `on: :create` was because now we'll have old invalid records in the db

Comment (peteryates) [reply app/models/bookings/placement_date.rb]: Having had another look, the reason that ` if: -> { bookings_school&.has_primary_and_secondary_phases? }` is required is so that when dates are created without the user picking a date we're now waiting until we're inside the `if ... valid?` condition before assigning the `supports_subjects` value
